app/src/main/java/ch/epfl/cs311/wanderwave/model/remote/ProfileConnection.kt

```diff
@@ -59,80 +59,77 @@ class ProfileConnection(
+            // a list of flow,
+            // then reduce the list of flow to a single flow that contains the list of tracks
+            // and then combine the two lists of tracks to update the profile
+            val chosenSongs = documentReferencesToFlows(chosenSongRefs, trackConnection)
+            val topSongs = documentReferencesToFlows(topSongRefs, trackConnection)
+            val bannedSongs = documentReferencesToFlows(bannedSongRefs, trackConnection)
+
+            val updatedProfile =
+                topSongs
+                    .combine(chosenSongs) { topSongs, chosenSongs -> Pair(topSongs, chosenSongs) }
+                    .combine(bannedSongs) { pair, bannedSongs ->
+                      profile.copy(
+                          topSongs = pair.first.getOrNull() ?: profile.topSongs,
+                          chosenSongs = pair.second.getOrNull() ?: profile.chosenSongs,
+                          bannedSongs = bannedSongs.getOrNull() ?: profile.bannedSongs)
                     }
 
-                // would like to keep the flow without collecting it, but I don't know how to do
-                // it...
-                updatedProfile
-                    .map { Result.success(it) }
-                    .collect { result ->
-                      result.onSuccess { profile -> trySend(Result.success(profile)) }
-                    }
-              }
-            } else {
-              trySend(
-                  Result.failure<Profile>(Exception("Songs lists have a Wrong Firebase Format")))
-            }
+            // would like to keep the flow without collecting it, but I don't know how to do
+            // it...
+            updatedProfile
+                .map { Result.success(it) }
+                .collect { result ->
+                  result.onSuccess { profile -> trySend(Result.success(profile)) }
+                }
           }
+          // } else {
+          //   Log.e("ProfileConnection", "Tracks have wrong firebase format")
+          //   trySend(Result.success(profile))
+          // }
         }
         awaitClose {}
       }
+
+  private fun documentReferencesToFlows(
+      documentReferences: List<DocumentReference>?,
+      trackConnection: TrackConnection
+  ): Flow<Result<List<Track>>> {
+    return documentReferences
+        // map to a list of flow
+        ?.map { trackRef -> trackConnection.fetchTrack(trackRef) }
+        // Extract the track from Result or return null if it's a failure
+        ?.map { flow -> flow.mapNotNull { result -> result.getOrNull() } }
+        // map to a list of track
+        ?.fold(flowOf(Result.success(listOf<Track>()))) { acc, track ->
+          acc.combine(track) { accTracks, track -> accTracks.map { tracks -> tracks + track } }
+        } ?: flowOf(Result.failure(Exception("Could not retrieve topSongs")))
+    // reduce the lists of flows to a
+    // single flows that contains the
+    // list of tracks
+  }
+
+  fun isValidObject(obj: Any?): Boolean {

```

Comment:
This function name is not very intuitive, it seems to check that something is a list, but mentions object in the name. Renaming or some documenting comments could help.
Also, maybe just returning the correctly typed value or `null` if not might be nice.

Reply:
Very nice idea !